bitbots_vision/bitbots_vision/vision_modules/yoeo_handler_utils.py

```diff
@@ -0,0 +1,462 @@
+        self._get_preprocessor_info()
+        detections = self._perform_nms(detections[0, ...])
+        detections = self._rescale_boxes(detections)
+
+        return detections
+
+    def _get_preprocessor_info(self) -> None:
+        preprocessor_info = self._image_preprocessor.get_info()
+        self._max_dim = preprocessor_info.max_dim
+        self._padding_top = preprocessor_info.padding_top
+        self._padding_bottom = preprocessor_info.padding_bottom
+        self._padding_left = preprocessor_info.padding_left
+        self._padding_right = preprocessor_info.padding_right
+
+    def _perform_nms(self, detections):
+        # shape prediction: (#boxes, >= 6)
+        # (>= 6)er meaning: (x, y, w, h, obj_conf, cond_class_conf_1, cond_class_conf_2, ...)
+
+        detections = self._preprocess_detections_for_nms(detections)
+
+        boxes, scores = self._get_boxes_and_scores_for_nms(detections)
+        indices = cv2.dnn.NMSBoxes(boxes, scores, 0, self._nms_thresh, top_k=self._nms_max_number_of_detections)
+
+        output = self._postprocess_nms_output(detections, indices)
+
+        return output
+
+    def _preprocess_detections_for_nms(self, detections):
+        detections = self._filter_by_objectness_confidence(detections)
+        detections = self._calculate_class_confidence_scores(detections)
+        detections = self._pin_down_last_dimension_to_6(detections)
+        #detections = self._filter_by_class_confidence(detections)
+        if self._too_many_boxes_remain_in(detections):
+            detections = self._keep_only_boxes_with_highest_objectness_confidence(detections)
+
+        return detections
+
+    def _filter_by_objectness_confidence(self, detections):
+        return detections[detections[:, 4] > self._conf_thresh]
+
+    def _calculate_class_confidence_scores(self, detections):
+        # class_confidence_score = conditional_class_probability * box_confidence_score (objectness)
+        # p(class, object)       = p(class | object)             * p(object)
+        detections[:, 5:] *= detections[:, 4:5]
+
+        return detections
+
+    def _pin_down_last_dimension_to_6(self, detections):
+        box_coordinates = detections[:, :4]
+        class_confidence_scores = detections[:, 5:]
+
+        if self._is_multilabel_prediction(class_confidence_scores):
+            i, j = (class_confidence_scores > self._conf_thresh).nonzero()
+            x = np.concatenate((box_coordinates[i], class_confidence_scores[i, j, None], j[:, None]), axis=1)
+        else:  # best class only
+            conf = np.max(class_confidence_scores, axis=1)
+            j = np.argmax(class_confidence_scores, axis=1)
+            x = np.concatenate((box_coordinates, conf[:, None], j[:, None]), axis=1)[conf > self._conf_thresh]
+
+        return x
+
+    @staticmethod
+    def _is_multilabel_prediction(class_confidence_scores) -> bool:
+        return class_confidence_scores.shape[1] > 1  # makes absolutely no sense
```

Comment:
What I mean by this comment is that multilabel prediction is triggered as soon as there is more than one class, whilst single class prediction is only active if there is just one class. In case of single class prediction, a max is performed over the classes which makes no sense if there is just one class anyway. This is how the pytorch implementation does it. I thought about having another parameter in the config, like "yoeo_single_class_pred: true / false"